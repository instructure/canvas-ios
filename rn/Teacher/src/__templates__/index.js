//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

export * from '../redux/__templates__/app-state'
export * from './account-notification'
export * from './addressBook'
export * from './api'
export * from './assignments'
export * from './attachment'
export * from './conversations'
export * from './course'
export * from './discussion'
export * from './enrollments'
export * from './error'
export * from './external-tool'
export * from './file'
export * from './grading-periods'
export * from './group'
export * from './helm'
export * from './launch-definitions'
export * from './mediaComment'
export * from './quiz'
export * from './quizSubmission'
export * from './role'
export * from './rubric'
export * from './section'
export * from './session'
export * from './submissions'
export * from './tab'
export * from './users'
export * from './help-links'
export * from './dashboard'
export * from '../modules/assignee-picker/__template__/Assignee'
