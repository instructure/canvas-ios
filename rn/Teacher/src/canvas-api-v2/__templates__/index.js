// @flow
export * from './submission'
export * from './user'
export * from './assignment'
export * from './course'
export * from './group'
export * from './section'
export * from './groupSet'

// QUERIES
export * from '../queries/__templates__/SubmissionList'
export * from '../queries/__templates__/ContextCard'
