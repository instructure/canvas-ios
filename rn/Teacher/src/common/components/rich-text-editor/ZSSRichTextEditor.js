//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* @flow */

import React, { Component } from 'react'
import {
  StyleSheet,
  WebView,
  NativeModules,
  Clipboard,
  I18nManager,
} from 'react-native'

import isEqual from 'lodash/isEqual'
import RNFS from 'react-native-fs'
import canvas from './../../../canvas-api'

const { NativeFileSystem } = NativeModules

type Props = {
  onLoad?: () => void,
  onFocus?: () => void,
  onBlur?: () => void,
  editorItemsChanged?: (items: string[]) => void,
  scrollEnabled?: boolean,
  navigator: Navigator,
  getFile: typeof canvas.getFile,
}

type State = {
  source: ?string,
  linkModalVisible: boolean,
  items: string[],
}

function extractFileID (string: string): ?string {
  // cant have spaces
  if (string.indexOf(' ') >= 0) {
    return null
  }

  // make sure it's a download url
  const urlMatcher = /^https:\/\/.*[^\s]\/files\/\d+\/download$/
  if (!string.match(urlMatcher)) {
    return null
  }

  const match = string.match(/files\/\d+\/download/)

  // $FlowFixMe - `match` is guaranteed to be defined because of previous check
  return match[0].split('/')[1]
}

export default class ZSSRichTextEditor extends Component<Props, State> {
  static defaultProps = {
    getFile: canvas.getFile,
  }

  webView: ?WebView
  showingLinkModal: boolean
  onHTML: ?((string) => void)

  state: State = {
    linkModalVisible: false,
    items: [],
    source: null,
  }

  async componentDidMount () {
    const path = await NativeFileSystem.pathForResource('zss-rich-text-editor', 'html')
    const source = await RNFS.readFile(`file://${path}`)
    this.setState({ source })
  }

  render () {
    return (
      <WebView
        source={{ html: this.state.source || '' }}
        ref={webView => { this.webView = webView }}
        onMessage={this._onMessage}
        onLoad={this._onLoad}
        scalesPageToFit={true}
        scrollEnabled={this.props.scrollEnabled === undefined || this.props.scrollEnabled}
        style={styles.editor}
        hideKeyboardAccessoryView={true}
        useWebKit
      />
    )
  }

  // PUBLIC ACTIONS

  setBold = () => {
    this.trigger('zss_editor.setBold();')
  }

  setItalic = () => {
    this.trigger('zss_editor.setItalic();')
  }

  insertLink = () => {
    this.trigger(`
      var selection = getSelection().toString();
      postMessage(JSON.stringify({type: 'INSERT_LINK', data: selection}));
    `)
  }

  prepareInsert = () => {
    this.trigger('zss_editor.prepareInsert();')
  }

  updateHTML = (html: ?string) => {
    const cleanHTML = this.escapeJSONString(html || '')
    this.trigger(`zss_editor.setHTML("${cleanHTML}");`)
  }

  setCustomCSS = (css?: string = '') => {
    css += `body { direction: ${I18nManager.isRTL ? 'rtl' : 'ltr'}; }`
    this.trigger(`zss_editor.setCustomCSS('${css}');`)
  }

  setUnorderedList = () => {
    this.trigger(`zss_editor.setUnorderedList();`)
  }

  setOrderedList = () => {
    this.trigger(`zss_editor.setOrderedList();`)
  }

  setTextColor = (color: string) => {
    this.prepareInsert()
    this.trigger(`zss_editor.setTextColor('${color}');`)
  }

  focusEditor = () => {
    this.trigger('zss_editor.focusEditor();')
  }

  blurEditor = () => {
    this.trigger('zss_editor.blurEditor();')
  }

  getHTML = async () => {
    const promise = new Promise((resolve, reject) => {
      this.onHTML = resolve
    })
    this.trigger(`zss_editor.postHTML();`)
    return promise
  }

  undo = () => {
    this.trigger('zss_editor.undo();')
  }

  redo = () => {
    this.trigger('zss_editor.redo();')
  }

  setContentHeight = (height: number) => {
    this.trigger(`zss_editor.contentHeight = ${height};`)
  }

  setPlaceholder = (placeholder: ?string) => {
    if (placeholder) {
      this.trigger(`zss_editor.setPlaceholder('${placeholder}');`)
    } else {
      this.trigger(`zss_editor.setPlaceholder(null);`)
    }
  }

  insertImage = (url: string) => {
    this.trigger(`zss_editor.insertImage('${url}');`)
  }

  insertVideoComment = (mediaID: string) => {
    this.trigger(`zss_editor.insertVideoComment('${mediaID}');`)
  }

  insertHTML = (html: string) => {
    this.trigger(`zss_editor.insertHTML('${html}');`)
  }

  trigger = async (js: string) => {
    if (!this.webView) return
    try {
      await this.webView.injectJavaScript(js)
    } catch (e) {}
  }

  // PRIVATE

  _onMessage = (event) => {
    const message = JSON.parse(event.nativeEvent.data)
    switch (message.type) {
      case 'CALLBACK':
        this._handleItemsCallback(message.data)
        break
      case 'LINK_TOUCHED':
        this._handleLinkTouched(message.data)
        break
      case 'INSERT_LINK':
        this._handleInsertLink(message.data)
        break
      case 'ZSS_LOADED':
        this._onZSSLoaded()
        break
      case 'EDITOR_FOCUSED':
        this._handleFocus()
        break
      case 'EDITOR_BLURRED':
        this._handleBlur()
        break
      case 'EDITOR_HTML':
        this._handleHTML(message.data)
        break
      case 'EDITOR_PASTE':
        this._handlePaste()
        break
    }
  }

  paste = (string: string) => {
    const formatted = this.escapeClipboardString(string)
    this.prepareInsert()
    this.insertHTML(formatted)
  }

  _handlePaste = async () => {
    let string = await Clipboard.getString()
    if (!string) return

    try {
      let fileID = extractFileID(string)
      if (fileID) {
        let { data } = await this.props.getFile(fileID)
        if (data.mime_class === 'image') {
          this.prepareInsert()
          this.insertImage(data.url)
        } else {
          this.paste(string)
        }
      } else {
        this.paste(string)
      }
    } catch (error) {
      this.paste(string)
    }
  }

  _handleItemsCallback = (items) => {
    if (isEqual(items, this.state.items)) {
      return
    }

    this.setState({ items })

    if (this.props.editorItemsChanged) {
      this.props.editorItemsChanged(items)
    }
  }

  _handleLinkTouched = (link) => {
    this.prepareInsert()
    this._showLinkModal(link.url, link.title)
  }

  _handleInsertLink = (selection) => {
    this.prepareInsert()
    this._showLinkModal(null, selection)
  }

  _handleHTML = (value) => {
    this.onHTML && this.onHTML(value)
  }

  _insertLink = (url, title) => {
    this.trigger(`zss_editor.insertLink("${url}", "${title}");`)
    this._hideLinkModal()
  }

  _updateLink = (url, title) => {
    this.trigger(`zss_editor.updateLink("${url}", "${title}");`)
    this._hideLinkModal()
  }

  _showLinkModal (url, title) {
    if (this.showingLinkModal) return
    this.showingLinkModal = true
    this.props.navigator.show(
      '/rich-text-editor/link',
      {
        modal: true,
        modalPresentationStyle: 'overCurrentContext',
        modalTransitionStyle: 'fade',
        embedInNavigationController: false,
      },
      {
        url: url,
        title: title,
        linkUpdated: this._updateLink,
        linkCreated: this._insertLink,
        onCancel: this._hideLinkModal,
      },
    )
  }

  _hideLinkModal = () => {
    this.props.navigator.dismiss()
    this.focusEditor()
    this.showingLinkModal = false
  }

  _onZSSLoaded = async () => {
    this.setCustomCSS()
    await this._setVideoPreviewImagePath()
    this.props.onLoad && this.props.onLoad()
  }

  _onLoad = () => {
    this._zssInit()
  }

  _zssInit () {
    this.trigger('zss_editor.init();')
  }

  _handleFocus () {
    if (this.props.onFocus) {
      this.props.onFocus()
    }
  }

  _handleBlur () {
    if (this.props.onBlur) {
      this.props.onBlur()
    }
  }

  _setVideoPreviewImagePath = async () => {
    const path = await NativeFileSystem.pathForResource('video-preview', 'png')
    await this.trigger(`zss_editor.videoPreviewImagePath = '${path}';`)
  }

  // UTILITIES

  escapeJSONString (string: string) {
    return string
      .replace(/[\\]/g, '\\\\')
      .replace(/["]/g, '\\"')
      .replace(/[/]/g, '\\/')
      .replace(/[\b]/g, '\\b')
      .replace(/[\f]/g, '\\f')
      .replace(/[\n]/g, '\\n')
      .replace(/[\r]/g, '\\r')
      .replace(/[\t]/g, '\\t')
  }

  escapeClipboardString (string: string) {
    // https://stackoverflow.com/a/9204218
    return string
      .replace(/[\\]/g, '\\\\')
      .replace(/["]/g, '\\"')
      .replace(/[']/g, "\\'")
      .replace(/[/]/g, '\\/')
      .replace(/[\b]/g, '\\b')
      .replace(/[\f]/g, '\\f')
      .replace(/[\n]/g, '<br>')
      .replace(/[\r]/g, '\\r')
      .replace(/[\t]/g, '\\t')
  }
}

const styles = StyleSheet.create({
  editor: {
    backgroundColor: 'transparent',
    flex: 1,
    flexDirection: 'column',
  },
})
