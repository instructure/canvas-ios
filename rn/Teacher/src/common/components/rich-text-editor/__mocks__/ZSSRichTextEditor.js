//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

const RealComponent = require.requireActual('../ZSSRichTextEditor')
const React = require('React')

export default class ZSSRichTextEditor extends React.Component<*> {
  render () {
    const props = { ...this.props, _setMock: this._setMock }
    return React.createElement('ZSSRichTextEditor', props, this.props.children)
  }

  _setMock = (ref: string, mock: any) => {
    this[ref] = mock
  }
}
ZSSRichTextEditor.propTypes = RealComponent.propTypes
