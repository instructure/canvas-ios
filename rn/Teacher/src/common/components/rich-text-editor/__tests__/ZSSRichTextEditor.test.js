//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import { shallow } from 'enzyme'
import { NativeModules, Clipboard, I18nManager } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import ZSSRichTextEditor from '../ZSSRichTextEditor'
import explore from '../../../../../test/helpers/explore'

import * as template from '../../../../__templates__'

jest
  .mock('react-native/Libraries/Components/ScrollView/ScrollView', () => 'ScrollView')
  .mock('react-native/Libraries/Components/Touchable/TouchableHighlight', () => 'TouchableHighlight')
  .mock('react-native/Libraries/Components/Button', () => 'Button')
  .mock('../LinkModal', () => 'LinkModal')
  .mock('../../CanvasWebView', () => 'CanvasWebView')
  .mock('react-native-fs', () => ({
    MainBundlePath: 'file:///mainBundle',
    readFile: jest.fn(() => Promise.resolve('<html></html>')),
  }))

describe('ZSSRichTextEditor', () => {
  let js
  beforeEach(() => {
    jest.clearAllMocks()

    js = jest.fn()
  })

  const options = {
    createNodeMock: (element) => {
      if (element.type === 'CanvasWebView') {
        return {
          evaluateJavaScript: js,
        }
      }
    },
  }

  const webView = (component) => {
    return explore(component.toJSON()).query(({ type }) => type === 'CanvasWebView')[0]
  }

  it('renders', () => {
    expect(
      renderer.create(
        <ZSSRichTextEditor />
      )
    ).toMatchSnapshot()
  })

  it('provides unique active editor items', () => {
    const items = jest.fn()
    const component = renderer.create(
      <ZSSRichTextEditor editorItemsChanged={items} />
    )
    const web = webView(component)
    postMessage(web, 'CALLBACK', ['link'])

    expect(items).toHaveBeenCalledWith(['link'])

    postMessage(web, 'CALLBACK', ['link'])
    expect(items).toHaveBeenCalledTimes(1)
  })

  it('gets html', async () => {
    const html = '<p>some html</p>'
    const view = shallow(<ZSSRichTextEditor />)
    const webView = view.find('CanvasWebView')
    const getHTML = view.instance().getHTML()
    webView.simulate('Message', { body: JSON.stringify({ type: 'EDITOR_HTML', data: html }) })
    const result = await getHTML
    expect(result).toEqual(html)
  })

  it('notifies when editor focused', () => {
    const onFocus = jest.fn()
    const component = renderer.create(
      <ZSSRichTextEditor onFocus={onFocus} />, options
    )

    const web = webView(component)
    postMessage(web, 'EDITOR_FOCUSED')

    expect(onFocus).toHaveBeenCalled()
    expect(js.mock.calls).toMatchSnapshot()
  })

  it('responds when zss editor loads', () => {
    const component = renderer.create(
      <ZSSRichTextEditor />, options
    )

    const web = webView(component)
    postMessage(web, 'ZSS_LOADED')

    expect(js.mock.calls).toMatchSnapshot()
  })

  it('loads with rtl direction when in rtl', () => {
    I18nManager.isRTL = true
    const component = renderer.create(
      <ZSSRichTextEditor />, options
    )

    const web = webView(component)
    postMessage(web, 'ZSS_LOADED')

    expect(js.mock.calls).toMatchSnapshot()
    I18nManager.isRTL = false
  })

  it('triggers undo', () => {
    testTrigger((editor) => editor.undo())
  })

  it('triggers redo', () => {
    testTrigger((editor) => editor.redo())
  })

  it('triggers bold', () => {
    testTrigger((editor) => editor.setBold())
  })

  it('triggers italic', () => {
    testTrigger((editor) => editor.setItalic())
  })

  it('triggers setPlaceholder', () => {
    testTrigger((editor) => editor.setPlaceholder('Add text'))
    testTrigger((editor) => editor.setPlaceholder(null))
  })

  it('triggers insertImage', () => {
    testTrigger((editor) => editor.insertImage('https://canvas.instructure.com/files/1/download'))
  })

  it('triggers insertVideoComment', () => {
    testTrigger((editor) => editor.insertVideoComment('1'))
  })

  it('shows link modal', () => {
    const navigator = template.navigator({ show: jest.fn() })
    const component = renderer.create(
      <ZSSRichTextEditor navigator={navigator} />, options
    )
    component.getInstance().insertLink()
    const web = webView(component)
    postMessage(web, 'INSERT_LINK')
    expect(navigator.show).toHaveBeenCalledWith(
      '/rich-text-editor/link',
      {
        modal: true,
        modalPresentationStyle: 'overCurrentContext',
        modalTransitionStyle: 'fade',
        embedInNavigationController: false,
      },
      {
        url: null,
        title: undefined,
        linkUpdated: expect.any(Function),
        linkCreated: expect.any(Function),
        onCancel: expect.any(Function),
      },
    )
  })

  it('shows link modal when link touched', () => {
    const navigator = template.navigator({ show: jest.fn() })
    const link = {
      url: 'http://test-update-link.com',
      title: 'test update link',
    }
    const component = renderer.create(
      <ZSSRichTextEditor navigator={navigator} />, options
    )
    component.getInstance().insertLink()
    const web = webView(component)
    postMessage(web, 'LINK_TOUCHED', link)
    expect(navigator.show).toHaveBeenCalledWith(
      '/rich-text-editor/link',
      {
        modal: true,
        modalPresentationStyle: 'overCurrentContext',
        modalTransitionStyle: 'fade',
        embedInNavigationController: false,
      },
      {
        url: 'http://test-update-link.com',
        title: 'test update link',
        linkUpdated: expect.any(Function),
        linkCreated: expect.any(Function),
        onCancel: expect.any(Function),
      },
    )
  })

  describe('link modal', () => {
    it('triggers insert new link', () => {
      const navigator = template.navigator({
        show: (route, options, props) => {
          props.linkCreated('url', 'title')
        },
      })
      const component = renderer.create(
        <ZSSRichTextEditor navigator={navigator} />, options
      )
      component.getInstance().insertLink()
      postMessage(webView(component), 'INSERT_LINK')

      expect(js.mock.calls).toMatchSnapshot()
    })

    it('triggers insert link with selection', () => {
      const navigator = template.navigator({ show: jest.fn() })
      const component = renderer.create(
        <ZSSRichTextEditor navigator={navigator} />, options
      )
      component.getInstance().insertLink()
      postMessage(webView(component), 'INSERT_LINK', 'selection')
      expect(navigator.show).toHaveBeenCalledWith(
        '/rich-text-editor/link',
        {
          modal: true,
          modalPresentationStyle: 'overCurrentContext',
          modalTransitionStyle: 'fade',
          embedInNavigationController: false,
        },
        {
          url: null,
          title: 'selection',
          linkUpdated: expect.any(Function),
          linkCreated: expect.any(Function),
          onCancel: expect.any(Function),
        },
      )
    })

    it('triggers update link', () => {
      const navigator = template.navigator({
        show: (route, options, props) => {
          props.linkUpdated('url', 'title')
        },
      })
      const component = renderer.create(
        <ZSSRichTextEditor navigator={navigator} />, options
      )
      component.getInstance().insertLink()
      postMessage(webView(component), 'INSERT_LINK')

      expect(js.mock.calls).toMatchSnapshot()
    })
  })

  it('triggers text color', () => {
    testTrigger((editor) => editor.setTextColor('white'))
  })

  it('triggers unordered list', () => {
    testTrigger((editor) => editor.setUnorderedList())
  })

  it('triggers ordered list', () => {
    testTrigger((editor) => editor.setOrderedList())
  })

  it('triggers focus', () => {
    testTrigger((editor) => editor.focusEditor())
  })

  it('triggers blur', () => {
    testTrigger((editor) => editor.blurEditor())
  })

  it('notifies when editor loaded', async () => {
    const pathPromise = Promise.resolve('/editor-loaded')
    NativeModules.NativeFileSystem.pathForResource = jest.fn(() => pathPromise)
    const onLoad = jest.fn()
    const screen = shallow(<ZSSRichTextEditor onLoad={onLoad} />)
    await pathPromise
    const webView = screen.find('CanvasWebView')

    const jsPromise = Promise.resolve()
    webView.getElement().ref({ evaluateJavaScript: jest.fn(() => jsPromise) })
    expect(onLoad).not.toHaveBeenCalled()
    webView.simulate('Message', { body: JSON.stringify({ type: 'ZSS_LOADED' }) })
    await new Promise((resolve, reject) => process.nextTick(resolve))
    expect(onLoad).toHaveBeenCalled()
  })

  it('sets feature flags', async () => {
    const screen = shallow(<ZSSRichTextEditor />)
    const webView = screen.find('CanvasWebView')

    let js = jest.fn()
    webView.getElement().ref({ evaluateJavaScript: js })
    screen.instance().setFeatureFlags(['one', 'two'])
    expect(js.mock.calls).toMatchSnapshot()
  })

  it('notifies when editor blurred', () => {
    const onBlur = jest.fn()
    const component = renderer.create(
      <ZSSRichTextEditor onBlur={onBlur} />
    )
    const web = webView(component)
    postMessage(web, 'EDITOR_BLURRED')
    expect(onBlur).toHaveBeenCalled()
  })

  it('updates html', () => {
    testTrigger((editor) => editor.updateHTML('<div>Hi</div>'))
    testTrigger((editor) => editor.updateHTML(null))
  })

  describe('paste', () => {
    async function paste (clipboard: string) {
      Clipboard.getString = jest.fn(() => Promise.resolve(clipboard))
      const screen = shallow(<ZSSRichTextEditor />)
      const webView = screen.find('CanvasWebView')
      webView.getElement().ref({ evaluateJavaScript: js })
      webView.simulate('Message', { body: JSON.stringify({ type: 'EDITOR_PASTE' }) })
      await new Promise((resolve, reject) => process.nextTick(resolve))
    }

    let js = jest.fn()
    beforeEach(() => {
      js.mockClear()
    })

    it('pastes plain text', async () => {
      await paste('plain text')
      expect(js.mock.calls).toMatchSnapshot()
    })

    it('replaces newlines with <br>', async () => {
      await paste('Line One\nLine Two')
      expect(js.mock.calls).toMatchSnapshot()
    })

    it('only pastes if clipboard has a string', async () => {
      await paste(null)
      expect(js).not.toHaveBeenCalled()
    })

    it('will not be tricked by urls with spaces', async () => {
      await paste('https://one.com/files/1/download https://two.com/files/2/download')
      expect(js.mock.calls).toMatchSnapshot()
    })

    it('only matches file download urls', async () => {
      const notAMatch = 'https://one.com/files/1'
      await paste(notAMatch)
      expect(js.mock.calls).toMatchSnapshot()
    })

    it('downloads and embeds image files', async () => {
      const image = template.file({
        mime_class: 'image',
        url: 'https://verified-download-url.jpg',
      })
      const fileURL = 'https://canvas.instructure.com/files/1/download'
      const getFile = jest.fn(() => Promise.resolve({ data: image }))
      Clipboard.getString = jest.fn(() => Promise.resolve(fileURL))
      const screen = shallow(<ZSSRichTextEditor getFile={getFile} />)
      const webView = screen.find('CanvasWebView')
      webView.getElement().ref({ evaluateJavaScript: js, setFeatureFlags: jest.fn() })
      webView.simulate('Message', { body: JSON.stringify({ type: 'EDITOR_PASTE' }) })
      await new Promise((resolve, reject) => process.nextTick(resolve))
      expect(js.mock.calls).toMatchSnapshot()
    })

    it('pastes link when file is not an image', async () => {
      const file = template.file({
        mime_class: 'file',
        url: 'https://verified-download-url.jpg',
      })
      const fileURL = 'https://canvas.instructure.com/files/1/download'
      const getFile = jest.fn(() => Promise.resolve({ data: file }))
      Clipboard.getString = jest.fn(() => Promise.resolve(fileURL))
      const screen = shallow(<ZSSRichTextEditor getFile={getFile} />)
      const webView = screen.find('CanvasWebView')
      webView.getElement().ref({ evaluateJavaScript: js, setFeatureFlags: jest.fn() })
      webView.simulate('Message', { body: JSON.stringify({ type: 'EDITOR_PASTE' }) })
      await new Promise((resolve, reject) => process.nextTick(resolve))
      expect(js.mock.calls).toMatchSnapshot()
    })

    it('catches error and pastes clipboard if network call fails', () => {
      expect(async () => {
        const fileURL = 'https://canvas.instructure.com/files/1/download'
        const getFile = jest.fn(() => Promise.reject('ERROR'))
        Clipboard.getString = jest.fn(() => Promise.resolve(fileURL))
        const screen = shallow(<ZSSRichTextEditor getFile={getFile} />)
        const webView = screen.find('CanvasWebView')
        webView.getElement().ref({ evaluateJavaScript: js })
        webView.simulate('Message', { body: JSON.stringify({ type: 'EDITOR_PASTE' }) })
        await new Promise((resolve, reject) => process.nextTick(resolve))
        expect(js.mock.calls).toMatchSnapshot()
      }).not.toThrow()
    })
  })

  function testTrigger (trigger: (editor: any) => void) {
    const screen = shallow(<ZSSRichTextEditor />)
    screen.find('CanvasWebView').getElement().ref({ evaluateJavaScript: js, setFeatureFlags: jest.fn() })
    trigger(screen.instance())
    expect(js.mock.calls).toMatchSnapshot()
  }

  function postMessage (webView: any, type: string, data: any) {
    const message = { type, data }
    const event = { body: JSON.stringify(message) }
    webView.props.onMessage(event)
  }
})
