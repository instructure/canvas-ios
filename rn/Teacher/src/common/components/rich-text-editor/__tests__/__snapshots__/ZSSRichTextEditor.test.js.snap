// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZSSRichTextEditor link modal triggers insert new link 1`] = `
Array [
  Array [
    "
      var selection = getSelection().toString();
      window.webkit.messageHandlers.canvas.postMessage(JSON.stringify({type: 'INSERT_LINK', data: selection}));
    ",
  ],
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertLink(\\"url\\", \\"title\\");",
  ],
  Array [
    "zss_editor.focusEditor();",
  ],
]
`;

exports[`ZSSRichTextEditor link modal triggers update link 1`] = `
Array [
  Array [
    "
      var selection = getSelection().toString();
      window.webkit.messageHandlers.canvas.postMessage(JSON.stringify({type: 'INSERT_LINK', data: selection}));
    ",
  ],
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.updateLink(\\"url\\", \\"title\\");",
  ],
  Array [
    "zss_editor.focusEditor();",
  ],
]
`;

exports[`ZSSRichTextEditor loads with rtl direction when in rtl 1`] = `
Array [
  Array [
    "zss_editor.setCustomCSS('body { direction: rtl; }');",
  ],
]
`;

exports[`ZSSRichTextEditor notifies when editor focused 1`] = `Array []`;

exports[`ZSSRichTextEditor paste downloads and embeds image files 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertImage('https://verified-download-url.jpg');",
  ],
]
`;

exports[`ZSSRichTextEditor paste only matches file download urls 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertHTML('https:\\\\/\\\\/one.com\\\\/files\\\\/1');",
  ],
]
`;

exports[`ZSSRichTextEditor paste pastes link when file is not an image 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertHTML('https:\\\\/\\\\/canvas.instructure.com\\\\/files\\\\/1\\\\/download');",
  ],
]
`;

exports[`ZSSRichTextEditor paste pastes plain text 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertHTML('plain text');",
  ],
]
`;

exports[`ZSSRichTextEditor paste replaces newlines with <br> 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertHTML('Line One<br>Line Two');",
  ],
]
`;

exports[`ZSSRichTextEditor paste will not be tricked by urls with spaces 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.insertHTML('https:\\\\/\\\\/one.com\\\\/files\\\\/1\\\\/download https:\\\\/\\\\/two.com\\\\/files\\\\/2\\\\/download');",
  ],
]
`;

exports[`ZSSRichTextEditor renders 1`] = `
<CanvasWebView
  hideKeyboardAccessoryView={true}
  onFinishedLoading={[Function]}
  onMessage={[Function]}
  scalesPageToFit={true}
  scrollEnabled={true}
  source={Object {}}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
      "flexDirection": "column",
    }
  }
/>
`;

exports[`ZSSRichTextEditor responds when zss editor loads 1`] = `
Array [
  Array [
    "zss_editor.setCustomCSS('body { direction: ltr; }');",
  ],
]
`;

exports[`ZSSRichTextEditor sets feature flags 1`] = `
Array [
  Array [
    "zss_editor.setFeatureFlags(\\"one,two\\")",
  ],
]
`;

exports[`ZSSRichTextEditor triggers blur 1`] = `
Array [
  Array [
    "zss_editor.blurEditor();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers bold 1`] = `
Array [
  Array [
    "zss_editor.setBold();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers focus 1`] = `
Array [
  Array [
    "zss_editor.focusEditor();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers insertImage 1`] = `
Array [
  Array [
    "zss_editor.insertImage('https://canvas.instructure.com/files/1/download');",
  ],
]
`;

exports[`ZSSRichTextEditor triggers insertVideoComment 1`] = `
Array [
  Array [
    "zss_editor.insertVideoComment('1');",
  ],
]
`;

exports[`ZSSRichTextEditor triggers italic 1`] = `
Array [
  Array [
    "zss_editor.setItalic();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers ordered list 1`] = `
Array [
  Array [
    "zss_editor.setOrderedList();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers redo 1`] = `
Array [
  Array [
    "zss_editor.redo();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers setPlaceholder 1`] = `
Array [
  Array [
    "zss_editor.setPlaceholder('Add text');",
  ],
]
`;

exports[`ZSSRichTextEditor triggers setPlaceholder 2`] = `
Array [
  Array [
    "zss_editor.setPlaceholder('Add text');",
  ],
  Array [
    "zss_editor.setPlaceholder(null);",
  ],
]
`;

exports[`ZSSRichTextEditor triggers text color 1`] = `
Array [
  Array [
    "zss_editor.prepareInsert();",
  ],
  Array [
    "zss_editor.setTextColor('white');",
  ],
]
`;

exports[`ZSSRichTextEditor triggers undo 1`] = `
Array [
  Array [
    "zss_editor.undo();",
  ],
]
`;

exports[`ZSSRichTextEditor triggers unordered list 1`] = `
Array [
  Array [
    "zss_editor.setUnorderedList();",
  ],
]
`;

exports[`ZSSRichTextEditor updates html 1`] = `
Array [
  Array [
    "zss_editor.setHTML(\\"<div>Hi<\\\\/div>\\");",
  ],
]
`;

exports[`ZSSRichTextEditor updates html 2`] = `
Array [
  Array [
    "zss_editor.setHTML(\\"<div>Hi<\\\\/div>\\");",
  ],
  Array [
    "zss_editor.setHTML(\\"\\");",
  ],
]
`;
